File: common/java/src/main/java/org/aiddl/common/planning/state_variable/data/CausalGraphCreator.java
package org.aiddl.common.planning.state_variable.data;

import java.util.HashSet;
import java.util.Set;

import org.aiddl.common.math.graph.Graph;
import org.aiddl.common.math.graph.GraphTerm;
import org.aiddl.common.math.graph.GraphTools;
import org.aiddl.common.math.graph.StronglyConnectedComponents;
import org.aiddl.common.planning.state_variable.Operator;
import org.aiddl.core.interfaces.Function;
import org.aiddl.core.interfaces.InterfaceImplementation;
import org.aiddl.core.representation.ListTerm;
import org.aiddl.core.representation.SetTerm;
import org.aiddl.core.representation.SymbolicTerm;
import org.aiddl.core.representation.Term;
import org.aiddl.core.tools.LockableList;
import org.aiddl.core.tools.LockableSet;

public class CausalGraphCreator implements Function, InterfaceImplementation {
	private final static SymbolicTerm InterfaceUri = Term.sym("org.aiddl.common.planning.state-variable.data.causal-graph-creator");
	
	@Override
	public SymbolicTerm getInterfaceUri() {
		return InterfaceUri;
	}

	@Override
	public Term apply(Term O) {
		LockableSet V = new LockableSet();
		LockableList E = new LockableList();
		
		for ( Term o_term : O.asCollection() ) {
			Operator o = new Operator(o_term.asTuple());
//			CollectionTerm effects = o.get(PlanningTerm.Effects).asCollection();
//			CollectionTerm preconditions = o.get(PlanningTerm.Preconditions).asCollection();
			
			for ( Term e1 : o.getEffects() ) {
//				if ( e1 instanceof TupleTerm ) {
//					e1 = e1.getOrDefault(PlanningTerm.StateVariableAssignment, e1);
//				}
				V.add(e1.getKey());
				for ( Term e2 : o.getEffects() ) {
					if ( !e1.getKey().equals(e2.getKey()) ) {
						E.add(Term.tuple(e1.getKey(), e2.getKey()));
					}
				}
				for ( Term p : o.getPreconditions() ) {
					V.add(p.getKey());
					E.add(Term.tuple(p.getKey(), e1.getKey()));
				}
			}
			
		}
		
		Graph G = new Graph( GraphTools.assembleGraph(Term.set(V), Term.set(E)) );
//		DirectedGraph2Dot g2d = new DirectedGraph2Dot();
//		g2d.setExportFilename("causal-graph.dot");
//		g2d.compute(G.getGraphTerm());
		
		StronglyConnectedComponents sccComp = new StronglyConnectedComponents();
		SetTerm SCCs = sccComp.compute(G).asSet();
		
		LockableSet cycle_free_edges = new LockableSet(); 
	    cycle_free_edges.addAll( G.getGraphTerm().get(GraphTerm.Edges).getSetCopy() );
		
		for ( Term SCC : SCCs ) {
			
			ListTerm order = totalOrder(SCC.asSet(), G);
			
			for ( Term v1 : SCC.asSet() ) {
				for ( Term e : G.getOutEdges(v1) ) {
					Term v2 = e.get(1);
					if ( SCC.asSet().contains(v2) ) {
						if ( order.indexOf(v1) >= order.indexOf(v2) ) {
							cycle_free_edges.remove(e);
						}
					}
				}
			}
		}
		
		Term G_causal = GraphTools.assembleGraph(Term.set(V), Term.set(cycle_free_edges));
		
//		g2d.setExportFilename("causal-graph-cycle-free.dot");
//		g2d.compute(G_causal);
		
		return G_causal;
	}
	
	private ListTerm totalOrder( SetTerm SCC, Graph G ) {
		LockableList order = new LockableList();

		Set<Term> ignored_edges = new HashSet<>();
		while ( order.size() != SCC.size() ) {
			Integer min = null;
			Term argMin = null;
			
			for ( Term u : SCC ) {
				if ( !order.contains(u)  ) {
					int in_counter = 0;
					
					for ( Term e : G.getInEdges(u) ) {
						if ( !ignored_edges.contains(e) ) {
							Term source = e.get(0);
					
							if ( !SCC.contains(source) ) {
								ignored_edges.add(e);
							} else {
								in_counter++;
							}
						}
					}
					if ( min == null || in_counter < min ) {
						min = in_counter;
						argMin = u;
					}
				}
			}
			order.add(argMin);
			ignored_edges.addAll(G.getIncidentEdges(argMin).getLockedList());
		}
		return Term.list(order);
	}
}

File: common/scala/src/main/scala/org/aiddl/common/scala/planning/state_variable/data/CausalGraphCreator.scala
package org.aiddl.common.scala.planning.state_variable.data

import scala.collection.immutable.HashSet
import scala.collection.mutable.HashMap
import scala.collection.mutable

import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.function.InterfaceImplementation
import org.aiddl.core.scala.representation._

import org.aiddl.common.scala.math.graph.{Graph, GraphTools, AdjacencyListGraph, StronglyConnectedComponentExtractor}
import org.aiddl.common.scala.planning.PlanningTerm.{Preconditions, Effects}

import Term.given_Conversion_Term_KeyVal
import scala.language.implicitConversions

class CausalGraphCreator extends Function with InterfaceImplementation {
    val interfaceUri = Sym("org.aiddl.common.planning.state-variable.data.causal-graph-creator");
    val fScc = new StronglyConnectedComponentExtractor

    def apply( os: Term ): Term = {
        var nodes = new HashSet[Term]
        var edges = new HashSet[Term]

        os.asCol.foreach( o => {
            o(Effects).asCol.foreach( e1 => {
                nodes = nodes + e1.key
                o(Effects).asCol.withFilter(e2 => e1.key != e2.key).foreach( e2 => {
                    nodes = nodes + e2.key
                    edges = edges + Tuple(e1.key, e2.key) })
                o(Preconditions).asCol.withFilter(p => e1.key != p.key).foreach( p => {
                    nodes = nodes + p.key
                    edges = edges + Tuple(p.key, e1.key) }) }) })

        val g = AdjacencyListGraph(SetTerm(nodes), SetTerm(edges))
        val sccs = fScc(g)
        val cycleFreeEdges: Set[Term] = sccs.asCol.flatMap( scc => {
            val order = totalOrder(scc.asCol, g, HashSet.empty, scc.asSet.set)  
          
            scc.asCol.flatMap( u =>
                g.outNeighbors(u).withFilter( v => {
                    !scc.asCol.contains(v) || order.indexOf(u) < order.indexOf(v) 
                } )
                .map(v => Tuple(u, v)) )
        }).toSet

        GraphTools.assembleGraph(SetTerm(nodes), SetTerm(cycleFreeEdges))
    }

    //private def totalOrder( scc: CollectionTerm, g: Graph ): Seq[Term] = 
    //   scc.asCol.toList.sortBy( u => g.inNeighbors(u).count( v => scc.contains(v) ) )

    private def totalOrder( scc: CollectionTerm, g: Graph, ignored: HashSet[Term], todo: Set[Term] ): List[Term] = {
        if (todo.isEmpty) Nil else {
            val argMin = todo.minBy( u => g.inNeighbors(u).count( v => scc.contains(v) && !ignored.contains(Tuple(u, v)) ) )
            argMin :: totalOrder(scc, g, ignored + g.incidentEdges(argMin), todo - argMin )
        }
    }
}